typescript/infra/scripts/helloworld/kathy.ts

```diff
@@ -54,48 +52,71 @@ async function main() {
     throw new Error(`Invalid chains to skip ${invalidChains}`);
   }
 
-  let failureOccurred = false;
-
   const origins = chains.filter((chain) => !skip || !skip.includes(chain));
-
-  // submit frequently so we don't have to wait a super long time for info to get into the metrics
-  const metricsInterval = setInterval(() => {
-    submitMetrics(metricsRegister, 'kathy', { appendMode: true }).catch(
-      console.error,
-    );
-  }, 1000 * 30);
-
-  for (const origin of origins) {
-    for (const destination of origins.filter((d) => d !== origin)) {
-      const labels = {
-        origin,
-        remote: destination,
-        ...constMetricLabels,
-      };
-      try {
-        await sendMessage(app, origin, destination, gasCalc);
-        messagesSendStatus.labels({ ...labels }).set(1);
-      } catch (err) {
-        console.error(
-          `Error sending message from ${origin} to ${destination}, continuing...`,
-          `${err}`.replaceAll('\n', ' ## '),
-        );
-        failureOccurred = true;
-        messagesSendStatus.labels({ ...labels }).set(0);
-      }
-
-      // Sleep 500ms to avoid race conditions where nonces are reused
-      await sleep(500);
-    }
+  const pairings = diagonalize(
+    origins.map((source) =>
+      origins.map((destination) =>
+        source == destination ? null : { source, destination },
+      ),
+    ),
+  )
+    .filter((v) => v !== null)
+    .map((v) => v!);
+
+  // default to once every 6 hours getting through all pairs
+  const fullCycleTime = process.env['KATHY_FULL_CYCLE_TIME']
+    ? parseInt(process.env['KATHY_FULL_CYCLE_TIME'])
+    : 1000 * 60 * 60 * 6;
+  if (!Number.isSafeInteger(fullCycleTime) || fullCycleTime <= 0) {
+    error('Invalid cycle time provided');
+    process.exit(1);
   }
 
-  clearInterval(metricsInterval);
-  // do not use append mode here so we can clear any old pairings we no longer care about.
-  await submitMetrics(metricsRegister, 'kathy', { appendMode: false });
+  // track how many we are still allowed to send in case some messages send slower than expected.
+  let allowedToSend = 0;
+  setInterval(() => {
+    allowedToSend++;
+  }, fullCycleTime / pairings.length);
+
+  for (
+    // in case we are restarting kathy, keep it from always running the exact same messages first
+    let currentPairingIndex = Date.now() % pairings.length;
+    ;
+    currentPairingIndex = (currentPairingIndex + 1) % pairings.length
+  ) {
+    currentPairingIndexGauge.set(currentPairingIndex);
+    // wait until we are allowed to send the message
+    while (allowedToSend <= 0) await sleep(1000);
+    allowedToSend--;
+
+    const { source, destination } = pairings[currentPairingIndex];
+    const labels = {
+      origin: source,
+      remote: destination,
+    };
+    try {
+      await sendMessage(app, source, destination, gasCalc);
+      log('Message sent successfully', { from: source, to: destination });
```

Comment:
3 different names used here for the same thing
can we be consistent with either 
1. from, to
2. origin, remote
3. source, destination

Reply:
Well, `from` and `to` are to conform with the ticket. `origin` and `remote` are the terms in the metrics, and `source` and `destination` are what the code was using.

I think I will try and make as much `origin` and `destination` since that seems to be the direction we are going but it won't be perfect.

Reply:
imo `from` and `to` are quite overloaded already in transaction schema
when I search for from in structured logs I would expect to find addresses